pyrokinetics/gk_code/GKInputReaderGS2.py

```diff
@@ -0,0 +1,244 @@
+import numpy as np
+import f90nml
+from cleverdict import CleverDict
+from copy import copy
+
+from ..typing import PathLike
+from ..constants import pi, sqrt2
+from ..local_species import LocalSpecies
+from ..local_geometry import LocalGeometry, LocalGeometryMiller, default_miller_inputs
+from ..numerics import Numerics
+from .GKInputReader import GKInputReader
+from .gs2_utils import pyro_gs2_miller, pyro_gs2_species, gs2_is_nonlinear
+
+
+class GKInputReaderGS2(GKInputReader):
+    """
+    Class that can read GS2 input files, and produce
+    Numerics, LocalSpecies, and LocalGeometry objects
+    """
+
+    code_name = "GS2"
+
+    def read(self, filename: PathLike) -> f90nml.Namelist:
+        """
+        Reads GS2 input file into a dictionary
+        """
+        super().read(filename)
+        # TODO These lines correct a possible mistake in the input file. Why is
+        #      this mistake corrected in particular? Should these corrections be
+        #      expanded?
+        if self.is_nonlinear() and "wstar_units" in self.data["knobs"]:
+            self.data["knobs"].pop("wstar_units")
+        return self.data
+
+    def verify(self, filename: PathLike):
+        """
+        Ensure this file is a valid gs2 input file"
+        """
+        data = f90nml.read(filename).todict()
+        expected_keys = ["kt_grids_knobs", "theta_grid_knobs", "theta_grid_eik_knobs"]
+        if not np.all(np.isin(expected_keys, list(data))):
+            raise ValueError(f"Expected GS2 file, received {filename}")
+
+    def is_nonlinear(self) -> bool:
+        return gs2_is_nonlinear(self.data)
+
+    def add_flags(self, flags) -> None:
+        """
+        Add extra flags to GS2 input file
+        """
+        for key, parameter in flags.items():
+            if key not in self.data:
+                self.data[key] = dict()
+            for param, val in parameter.items():
+                self.data[key][param] = val
+
+    def get_local_geometry(self) -> LocalGeometry:
+        """
+        Returns local geometry. Delegates to more specific functions
+        """
+        gs2_eq = self.data["theta_grid_knobs"]["equilibrium_option"]
+        if gs2_eq in ["eik", "default"]:
+            local_eq = self.data["theta_grid_eik_knobs"].get("local_eq", True)
+            iflux = self.data["theta_grid_eik_knobs"].get("iflux", 0)
+            if local_eq:
+                if iflux == 0:
+                    return self.get_local_geometry_miller()
+                else:
+                    # return self.get_fourier()
+                    raise NotImplementedError("GS2 Fourier options are not implemented")
+            else:
+                raise RuntimeError("GS2 is not using local equilibrium")
+        else:
+            raise NotImplementedError(
+                f"GS2 equilibrium option {gs2_eq} not implemented"
+            )
```

Comment:
I would invert most of these to have the preconditions up front:

```suggestion
        if gs2_eq not in ["eik", "default"]:
            raise NotImplementedError(f"GS2 equilibrium option {gs2_eq} not implemented")

        local_eq = self.data["theta_grid_eik_knobs"].get("local_eq", True)
        if not local_eq:
            raise RuntimeError("GS2 is not using local equilibrium")

        iflux = self.data["theta_grid_eik_knobs"].get("iflux", 0)
        if iflux != 0:
            # return self.get_fourier()
            raise NotImplementedError("GS2 Fourier options are not implemented")

        return self.get_local_geometry_miller()
```

Reply:
Using `local_eq = T` and `iflux /= 0` is an error in GS2 (or at least it's not correct). The Fourier representation is actually activated by `local_eq=T`. If you want to halt if we're using a non-Miller equilibrium then you would need to test `geoType` (https://gyrokinetics.gitlab.io/gs2/page/namelists/index.html#theta_grid_parameters-geotype) as well as `local_eq` I think.

Reply:
Yes that check should be with `geoType`, looks like a bug! We do need `iflux = 0` as well but that's separate to checking if this is Miller or Fourier

Reply:
Thanks for this. Just to check I'm understanding correctly, would the following work without touching `iflux`?
```python
if gs2_eq not in ["eik", "default"]:
    raise NotImplementedError(f"GS2 equilibrium option {gs2_eq} not implemented")

local_eq = self.data["theta_grid_eik_knobs"].get("local_eq", True)
if not local_eq:
    raise RuntimeError("GS2 is not using local equilibrium")

geotype = self.data["theta_grid_parameters"].get("geotype", 0)
if geotype != 0:
    raise NotImplementedError("GS2 Fourier options are not implemented")
```

Reply:
The key is `geoType`. Not sure if this is case sensitive.

Otherwise yes, though I would also force `iflux=0` after checking if it is a local equilibrium